SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriverRPC.cpp

```diff
@@ -0,0 +1,471 @@
+	if (!ActorReceivingRPC.IsValid())
+	{
+		UE_LOG(LogSpatialRPCService, Log,
+			   TEXT("Entity receiving ring buffer RPC does not exist in PackageMap, possibly due to corresponding actor getting "
+					"destroyed. Entity: %lld"),
+			   EntityId);
+		return false;
+	}
+	return true;
+}
+
+bool USpatialNetDriverRPC::CanExtractRPCOnServer(Worker_EntityId EntityId)
+{
+	const TWeakObjectPtr<UObject> ActorReceivingRPC = NetDriver->PackageMap->GetObjectFromEntityId(EntityId);
+	if (!ActorReceivingRPC.IsValid())
+	{
+		UE_LOG(LogSpatialRPCService, Log,
+			   TEXT("Entity receiving ring buffer RPC does not exist in PackageMap, possibly due to corresponding actor getting "
+					"destroyed. Entity: %lld"),
+			   EntityId);
+		return false;
+	}
+
+	const bool bActorRoleIsSimulatedProxy = Cast<AActor>(ActorReceivingRPC.Get())->Role == ROLE_SimulatedProxy;
+	if (bActorRoleIsSimulatedProxy)
+	{
+		UE_LOG(LogSpatialRPCService, Verbose,
+			   TEXT("Will not process server RPC, Actor role changed to SimulatedProxy. This happens on migration. Entity: %lld"),
+			   EntityId);
+		return false;
+	}
+	return true;
+}
+
+struct RAIIParamsHolder
+{
+	RAIIParamsHolder(UFunction* InFunction, uint8* Memory)
+		: Function(InFunction)
+	{
+		Parms = Memory;
+		FMemory::Memzero(Parms, Function->ParmsSize);
+	}
+
+	~RAIIParamsHolder()
+	{
+		// Destroy the parameters.
+		// warning: highly dependent on UObject::ProcessEvent freeing of parms!
+		for (TFieldIterator<GDK_PROPERTY(Property)> It(Function); It && It->HasAnyPropertyFlags(CPF_Parm); ++It)
+		{
+			It->DestroyValue_InContainer(Parms);
+		}
+	}
+
+	UFunction* Function;
+	uint8* Parms;
+};
+
+bool USpatialNetDriverRPC::ApplyRPC(Worker_EntityId EntityId, SpatialGDK::ReceivedRPC RPCData, const FRPCMetaData& MetaData)
+{
+	constexpr bool RPCConsumed = true;
+
+	FUnrealObjectRef ObjectRef(EntityId, RPCData.Offset);
+	TWeakObjectPtr<UObject> TargetObjectWeakPtr = NetDriver->PackageMap->GetObjectFromUnrealObjectRef(ObjectRef);
+	UObject* TargetObject = TargetObjectWeakPtr.Get();
+
+	const FClassInfo& ClassInfo = NetDriver->ClassInfoManager->GetOrCreateClassInfoByObject(TargetObjectWeakPtr.Get());
+	UFunction* Function = ClassInfo.RPCs[RPCData.Index];
+	if (Function == nullptr)
+	{
+		// TODO, log something
+		// return FRPCErrorInfo{ TargetObject, nullptr, ERPCResult::MissingFunctionInfo, ERPCQueueProcessResult::ContinueProcessing };
+		return RPCConsumed;
+	}
+
+	// NB : FMemory_Alloca is stack allocation, and cannot be done inside the RAII holder.
+	uint8* Parms = (uint8*)FMemory_Alloca(Function->ParmsSize);
+	RAIIParamsHolder ParamAlloc(Function, Parms);
+
+	TSet<FUnrealObjectRef> UnresolvedRefs;
+	{
+		// Scope for the SpatialNetBitReader lifecycle (only one should exist per thread).
+		TSet<FUnrealObjectRef> MappedRefs;
+
+		FSpatialNetBitReader PayloadReader(NetDriver->PackageMap, const_cast<uint8*>(RPCData.PayloadData.GetData()),
+										   RPCData.PayloadData.Num() * 8, MappedRefs, UnresolvedRefs);
```

Comment:
Why '8'?

Reply:
The reader takes a size in bits instead of bytes.

Reply:
Yup, maybe move to suitably named local constexpr?